File: dotnet/Allmark/Render/CodeBlockRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class CodeBlockRenderer
{
	public static Renderer Create()
	{
		return new Renderer
		{
			Name = "code_block",
			Render = Render,
		};
	}

	public static void Render(MarkdownNode node, RendererState state, bool? first = null, bool? last = null, bool? decode = true)
	{
		if (node.Type == "code_block" && node.Content != null && node.Content.Length == 0)
		{
			return;
		}

		RenderUtils.StartNewLine(node, state);
		var lang = "";
		if (!string.IsNullOrEmpty(node.Info))
		{
			var trimmed = node.Info.Trim().Split(' ')[0];
			if (!string.IsNullOrEmpty(trimmed))
			{
				lang = $" class=\"language-{Utils.EscapeHtml(trimmed)}\"";
			}
		}
		state.Output.Append($"<pre><code{lang}>");
		RenderChildren.Execute(node, state, false);
		state.Output.Append("</code></pre>");
		RenderUtils.EndNewLine(node, state);
	}
}

File: web/src/render/codeBlockRenderer.ts
import type MarkdownNode from "../types/MarkdownNode";
import type Renderer from "../types/Renderer";
import type RendererState from "../types/RendererState";
import escapeHtml from "../utils/escapeHtml";
import renderChildren from "./renderChildren";
import { endNewLine, startNewLine } from "./renderUtils";

const renderer: Renderer = {
	name: "code_block",
	render,
};
export default renderer;

export function render(node: MarkdownNode, state: RendererState): void {
	// HACK: Probably shouldn't even be in the tree by this point...
	if (node.type === "code_block" && !node.content.length) {
		return;
	}

	startNewLine(node, state);
	let lang = "";
	if (node.info) {
		let trimmed = node.info.trim().split(" ")[0];
		if (trimmed) {
			lang = ` class="language-${escapeHtml(trimmed)}"`;
		}
	}
	state.output += `<pre><code${lang}>`;
	renderChildren(node, state, false);
	state.output += "</code></pre>";
	endNewLine(node, state);
}
